File: src/main/java/net/lyragames/practice/utils/item/CustomItemStack.java
package net.lyragames.practice.utils.item;

import lombok.Data;
import lombok.Getter;
import org.bukkit.event.player.PlayerInteractEvent;
import org.bukkit.inventory.ItemStack;

import java.util.LinkedList;
import java.util.List;
import java.util.UUID;
import java.util.function.Consumer;

@Data
public class CustomItemStack {

    @Getter
    private static List<CustomItemStack> customItemStacks = new LinkedList<>();

    private final UUID uuid;
    private ItemStack itemStack;

    private boolean rightClick;
    private boolean removeOnClick;

    private Consumer<PlayerInteractEvent> clicked;

    public CustomItemStack(UUID uuid, ItemStack itemStack) {
        this.uuid = uuid;
        this.itemStack = itemStack;
    }

    public void create() {
        customItemStacks.add(this);
    }

    public static CustomItemStack getByInteraction(PlayerInteractEvent event) {
        return customItemStacks.stream().filter(customItemStack -> customItemStack.getUuid().equals(event.getPlayer().getUniqueId())
                && customItemStack.getItemStack().isSimilar(event.getItem())).findFirst().orElse(null);
    }
}

File: src/main/kotlin/net/lyragames/practice/utils/item/CustomItemStack.kt
package net.lyragames.practice.utils.item

import org.bukkit.event.player.PlayerInteractEvent
import org.bukkit.inventory.ItemStack
import java.util.*
import java.util.function.Consumer

class CustomItemStack(val uuid: UUID, val itemStack: ItemStack) {
    var rightClick = false
    var removeOnClick = false
    var clicked: Consumer<PlayerInteractEvent>? = null

    fun create() {
        customItemStacks.add(this)
    }

    companion object {
        val customItemStacks: MutableList<CustomItemStack> = mutableListOf()

        @JvmStatic
        fun getByInteraction(event: PlayerInteractEvent): CustomItemStack? {
            return customItemStacks.firstOrNull {
                it.uuid == event.player.uniqueId && it.itemStack
                    .isSimilar(event.item)
            }
        }
    }
}